accepted/0000-workspaces.md

```diff
@@ -0,0 +1,479 @@
+
+Also available is the alternative of supporting only the install (or bootstrap as Lerna names it) aspect of this proposal, following a less feature-rich approach but one that would still enable the basic goal of improving the user experience of managing multiple child packages but from all the feedback collected during the research phase of this RFC, this alternative is much less desirable to the community of maintainers involved.
+
+## Implementation
+
+### 1. Make npm workspace-aware
+
+We're following the lead of Yarn in supporting the `workspaces` `package.json` field which defines a list of paths, each of these paths may be a workspace itself but it also support globs.
+
+`package.json` example:
+
+```
+{
+    "name": "workspace-example",
+    "version": "1.0.0",
+    "workspaces": {
+        "packages": [
+            "packages/*"
+        ]
+    }
+}
+```
+
+`package.json` shorthand example:
+
+```
+{
+    "name": "workspace-example",
+    "version": "1.0.0",
+    "workspaces": [
+        "packages/*"
+    ]
+}
+```
+
+
+### 2. Installing dependencies across child packages
+
+Change `npm install` ([arborist](https://www.npmjs.com/package/@npmcli/arborist)) behavior to make it install every nested package by default once a valid configuration is defined.
+
+Arborist should also be aware of all nested workspaces and correctly link to an internal workspace should it match the required semver version of an expected dependency anywhere in the installing tree. e.g:
+
+```
+// Given this package.json structure:
+├── package.json { "workspaces": ["dep-a", "dep-b"] }
+├── dep-a
+│   └── package.json { "dependencies": { "dep-b": "^1.0.0" } }
+└── dep-b
+    └── package.json { "version": "1.3.1" }
+
+$ npm install
+
+// Results in this symlinking structure:
+├── node_modules
+│   ├── dep-a -> ./dep-a
+│   └── dep-b -> ./dep-b
+├── dep-a
+│   └── node_modules
+│       └── dep-b -> ../../../node_modules/dep-b
+└── dep-b
+```
+
+NOTE: The final implementation of the symlinking structure might differ from the examples laid out here (which are trying to be as illustrative as possible). Arborist might end up resolving all symlinks up to the actual place instead of creating a chain of symlinks.
+
+### 3. Run commands across all child packages
+
+Make npm commands workspace-aware, so that running a command tries to run it within nested packages as long as a workspaces field is defined in `package.json`.
+
+Only a subset of npm commands are to be supported:
+
+- `fund` List funding info for all packages
+- `ls` List all packages
+- `outdated` List outdated deps for all packages
+- `publish` Publishes all packages
+- `run-script` Run arbitrary scripts for all packages
+- `rebuild` Rebuild all packages
+- `restart`
+- `start`
+- `stop`
+- `test` Run tests in all packages
+- `update` Update a dependency in all packages
+- `version` Bump versions for all packages
```

Comment:
Presumably a workspace-aware `npm version` would **not** include such fancy git-related magic that say, `lerna version` does? Because I can guarantee you a gigantic bug-farm, otherwise.

Reply:
Yeah I would agree that this kind of logic should be kept out of lower level tools like npm.  The problem here is that the line is fuzzy, because folks do want sane version bumping and tagging behavior.  Is there a good middle-ground you think we might get to @evocateur?

Reply:
> Presumably a workspace-aware `npm version` would **not** include such fancy git-related magic that say, `lerna version` does? Because I can guarantee you a gigantic bug-farm, otherwise.

yep 👍 100% the idea is to avoid all that crazyness, the `version` support would just be the equivalent of running `npm version <target>` inside each of the packages

Reply:
With many commits or just one? Same question for the tags.  I think this might already be getting into the craziness which @evocateur was worried about.

Reply:
just one commit/tag for all repos involved (mind the RFC also mentions the possibility of filtering a subset of packages) 😅 but IMO if we realize `version` and `publish` are too overwhelming to support in this initial implementation I'll be in favor of dropping support to it altogether - mind that getting the _installing_ part of the workspaces right is the core of this RFC

Reply:
IMO, a single tag makes it difficult to look up code for a specific version when the multi-repo uses independent versioning strategy.

Using the example from the top of this RFC, let's say `dep-a` will be released in version `1.2.3` and `dep-b` will be released as `2.1.0`. How do you propose to build a tag name for such release? How can a person looking for code of `dep-b@2.1.0` find which git tag to check out?

Reply:
I strongly feel that independent versioning should be the default (and ideally, required) - an option that encourages a package to bump a version unnecessarily, only to make it match another package's version, has not been good for users (when babel did it, for example)

Reply:
tbh as of now I'm really leaning on the side of just removing `version` and `publish` completely from the scope of this initial workspaces implementation, hopefully we can have a final decision during the deep dive call next week and I'll come back to document it here 😊

Reply:
as announced/discussed in the [OpenRFC Deep Dive call](https://github.com/npm/rfcs/issues/106), `version` and `publish` are not going to be supported in this initial **npm workspaces** implementation.